radicle-node/src/wire/protocol.rs

```diff
@@ -0,0 +1,315 @@
+use std::collections::VecDeque;
+use std::io::{Read, Write};
+use std::net;
+use std::os::unix::io::AsRawFd;
+use std::os::unix::prelude::RawFd;
+use std::sync::Arc;
+use std::time::Duration;
+
+use cyphernet::addr::{Addr as _, HostAddr, PeerAddr};
+use cyphernet::crypto::Ecdh;
+use nakamoto_net::{DisconnectReason, Link, LocalTime};
+use netservices::noise::NoiseXk;
+use netservices::wire::{ListenerEvent, NetAccept, NetTransport, SessionEvent};
+use netservices::{Frame, NetSession};
+use radicle::collections::HashMap;
+use radicle::crypto::Negotiator;
+use radicle::node::NodeId;
+use radicle::storage::WriteStorage;
+
+use crate::crypto::Signer;
+use crate::service::reactor::Io;
+use crate::service::{routing, session, Message};
+use crate::wire::{Decode, Encode};
+use crate::{address, service, wire};
+
+pub type Session<N> = NoiseXk<N>;
+
+impl Frame for Message {
+    type Error = wire::Error;
+
+    fn unmarshall(mut reader: impl Read) -> Result<Option<Self>, Self::Error> {
+        match Message::decode(&mut reader) {
+            Ok(msg) => Ok(Some(msg)),
+            Err(wire::Error::Io(_)) => Ok(None),
+            Err(err) => Err(err),
+        }
+    }
+
+    fn marshall(&self, mut writer: impl Write) -> Result<usize, Self::Error> {
+        self.encode(&mut writer).map_err(wire::Error::from)
+    }
+}
+
+type Queue<N> = VecDeque<reactor::Action<NetAccept<Session<N>>, NetTransport<Session<N>, Message>>>;
+
+pub struct Wire<R, S, W, G, N: Negotiator> {
+    inner: service::Service<R, S, W, G>,
+    inner_queue: Queue<N>,
```

Comment:
I don't understand what `inner_queue` is for: it doesn't seem to queue messages from the inner service as I originally thought. It seems like we push & pop it within this module :thinking:

Reply:
It is used when the Wire needs to produce an action for the `Reactor` due to its own internal logic. Nothing to do with the `Service`

Reply:
Ahh, gotcha. So it's kind of like a "short circuit" where reactor inputs are processed directly by Wire.

Reply:
Correct